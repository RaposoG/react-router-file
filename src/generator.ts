// src/generator.ts
import fs from 'fs';
import path from 'path';
import { glob } from 'glob';

// ðŸ‘‡ 1. ADICIONE A NOVA OPÃ‡ÃƒO AQUI
interface GeneratorOptions {
  pagesDir: string;
  outputFile: string;
  importSource: string; // Recebe a nova opÃ§Ã£o
}

// ... (as funÃ§Ãµes lazyImport e formatRoutePath nÃ£o mudam) ...
const lazyImport = (componentName: string, filePath: string, outputFile: string) => {
  const relativePath = path.relative(path.dirname(outputFile), filePath).replace(/\\/g, '/');
  return `const ${componentName} = lazy(() => import('./${relativePath}'));`;
};

function formatRoutePath(filePath: string, pagesDir: string): string {
  const relativePath = path.relative(pagesDir, filePath);
  const routePath = relativePath
    .replace(/\\/g, '/')
    .replace(/\.(tsx|jsx)$/, '')
    .replace(/\/index$/, '')
    .replace(/^index$/, '')
    .replace(/\[\.\.\.([^\]]+)\]/g, '*')
    .replace(/\[([^\]]+)\]/g, ':$1');
  return `/${routePath}`;
}

// ðŸ‘‡ 2. RECEBA A OPÃ‡ÃƒO NOS PARÃ‚METROS DA FUNÃ‡ÃƒO
export async function generateRoutes(options: GeneratorOptions) {
  const { pagesDir, outputFile, importSource } = options;
  console.log('ðŸ”„ [react-router-file] Gerando rotas...');
  
  const pageFiles = await glob('**/*.{tsx,jsx}', {
    cwd: pagesDir,
    absolute: true,
  });

  const imports: string[] = [];
  const routeElements: string[] = [];

  pageFiles.forEach((file, index) => {
    const componentName = `Page${index}`;
    imports.push(lazyImport(componentName, file, outputFile));
    routeElements.push(
      `<Route path="${formatRoutePath(file, pagesDir)}" element={<${componentName} />} />`
    );
  });
  
  routeElements.sort((a, b) => {
    if (a.includes('*')) return 1;
    if (b.includes('*')) return -1;
    return b.length - a.length;
  });

  const outputContent = `
// ATTENTION: This file is automatically generated by react-router-file.
// Do not edit manually.
import { lazy, Suspense } from 'react';
import { Routes, Route } from '${importSource}';

${imports.join('\n')}

const LoadingComponent = () => null;

export const AppRoutes = () => (
  <Suspense fallback={<LoadingComponent />}>
    <Routes>
      ${routeElements.join('\n      ')}
    </Routes>
  </Suspense>
);
  `;

  fs.mkdirSync(path.dirname(outputFile), { recursive: true });
  fs.writeFileSync(outputFile, outputContent.trim());
  console.log(`âœ… [react-router-file] Rotas geradas com sucesso!`);
}