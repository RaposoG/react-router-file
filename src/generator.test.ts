import { describe, it, expect, vi, beforeEach } from 'vitest';
import { vol } from 'memfs';
import { generateRoutes } from './generator.js';
import { glob } from 'glob';
import fs from 'fs';

// Mocks
vi.mock('glob');
vi.mock('fs', async () => {
  const memfs = await vi.importActual<typeof import('memfs')>('memfs');
  return {
    ...memfs,
    default: {
      ...memfs.fs,
      mkdirSync: memfs.vol.mkdirSync.bind(memfs.vol),
    },
    mkdirSync: memfs.vol.mkdirSync.bind(memfs.vol),
  };
});

// Helper para silenciar o console durante os testes
beforeEach(() => {
  vi.spyOn(console, 'log').mockImplementation(() => {});
});

describe('generateRoutes() - Core Features', () => {
  const PAGES_DIR = '/project/src/pages';
  const OUTPUT_FILE = '/project/src/router.tsx';
  const mockedGlob = vi.mocked(glob);

  beforeEach(() => {
    vol.reset();
    mockedGlob.mockClear();
  });

  it('should generate routes for basic pages and sort them correctly', async () => {
    mockedGlob.mockResolvedValue([
      `${PAGES_DIR}/index.tsx`,
      `${PAGES_DIR}/about.tsx`,
      `${PAGES_DIR}/contact.tsx`,
    ]);
    vol.fromJSON({
      'index.tsx': 'export default () => <div>Home</div>',
      'about.tsx': 'export default () => <div>About</div>',
      'contact.tsx': 'export default () => <div>Contact</div>',
    }, PAGES_DIR);

    await generateRoutes({ pagesDir: PAGES_DIR, outputFile: OUTPUT_FILE, importSource: 'react-router-dom' });
    const generatedCode = fs.readFileSync(OUTPUT_FILE, 'utf-8');

    expect(generatedCode).toContain('<Route path="contact" element={<Page2 />} />');
    expect(generatedCode).toContain('<Route path="about" element={<Page1 />} />');
    expect(generatedCode).toContain('<Route index element={<Page0 />} />');
  });

  // --- TESTE CORRIGIDO ---
  it('should handle dynamic and catch-all routes, sorting catch-all last', async () => {
    mockedGlob.mockResolvedValue([
      `${PAGES_DIR}/posts/[id].tsx`, // Page0
      `${PAGES_DIR}/docs/[...slug].tsx`, // Page1
      `${PAGES_DIR}/index.tsx`, // Page2
    ]);
    vol.fromJSON({
      'posts/[id].tsx': 'export default () => <div>Post</div>',
      'docs/[...slug].tsx': 'export default () => <div>Docs</div>',
      'index.tsx': 'export default () => <div>Home</div>',
    }, PAGES_DIR);

    await generateRoutes({ pagesDir: PAGES_DIR, outputFile: OUTPUT_FILE, importSource: 'react-router-dom' });
    const generatedCode = fs.readFileSync(OUTPUT_FILE, 'utf-8');
    
    // Verifica a estrutura aninhada e a ordem correta dos nós
    const postsRouteRegex = /<Route path="posts">\s*<Route path=":id" element={<Page0 \/>} \/>\s*<\/Route>/s;
    const docsRouteRegex = /<Route path="docs">\s*<Route path="\*" element={<Page1 \/>} \/>\s*<\/Route>/s;

    expect(generatedCode).toMatch(postsRouteRegex);
    expect(generatedCode).toMatch(docsRouteRegex);

    // A verificação final e mais importante: a posição da rota de posts DEVE ser anterior à de docs
    const postsIndex = generatedCode.search(postsRouteRegex);
    const docsIndex = generatedCode.search(docsRouteRegex);
    
    expect(postsIndex).toBeGreaterThan(-1);
    expect(docsIndex).toBeGreaterThan(-1);
    expect(postsIndex).toBeLessThan(docsIndex);
  });

  it('should use the custom importSource when provided', async () => {
    mockedGlob.mockResolvedValue([`${PAGES_DIR}/home.tsx`]);
    vol.fromJSON({ 'home.tsx': 'export default () => <div />' }, PAGES_DIR);

    await generateRoutes({ pagesDir: PAGES_DIR, outputFile: OUTPUT_FILE, importSource: 'react-router' });
    const generatedCode = fs.readFileSync(OUTPUT_FILE, 'utf-8');
    expect(generatedCode).toContain("from 'react-router'");
  });
});

describe('generateRoutes() - Layout Features', () => {
  const PAGES_DIR = '/project/src/pages';
  const OUTPUT_FILE = '/project/src/router.tsx';
  const mockedGlob = vi.mocked(glob);

  beforeEach(() => {
    vol.reset();
    mockedGlob.mockClear();
  });

  it('should create a root layout wrapping all pages', async () => {
    mockedGlob.mockResolvedValue([
      `${PAGES_DIR}/layout.tsx`,
      `${PAGES_DIR}/index.tsx`,
      `${PAGES_DIR}/about.tsx`,
    ]);
    vol.fromJSON({
      'layout.tsx': 'export default () => <Outlet />',
      'index.tsx': 'export default () => <div>Home</div>',
      'about.tsx': 'export default () => <div>About</div>',
    }, PAGES_DIR);

    await generateRoutes({ pagesDir: PAGES_DIR, outputFile: OUTPUT_FILE, importSource: 'react-router-dom' });
    const generatedCode = fs.readFileSync(OUTPUT_FILE, 'utf-8');
    
    const layoutWrapperRegex = /<Route path="\/" element={<Page0 \/>}>\s*<Route index element={<Page1 \/>} \/>\s*<Route path="about" element={<Page2 \/>} \/>\s*<\/Route>/s;
    expect(generatedCode).toMatch(layoutWrapperRegex);
  });

  it('should handle nested layouts correctly', async () => {
    mockedGlob.mockResolvedValue([
      `${PAGES_DIR}/layout.tsx`,
      `${PAGES_DIR}/dashboard/layout.tsx`,
      `${PAGES_DIR}/dashboard/index.tsx`,
      `${PAGES_DIR}/dashboard/settings.tsx`,
    ]);
    vol.fromJSON({
      'layout.tsx': 'export default () => <Outlet />', // Page0
      'dashboard/layout.tsx': 'export default () => <Outlet />', // Page1
      'dashboard/index.tsx': 'export default () => <div>Dashboard Home</div>', // Page2
      'dashboard/settings.tsx': 'export default () => <div>Settings</div>', // Page3
    }, PAGES_DIR);

    await generateRoutes({ pagesDir: PAGES_DIR, outputFile: OUTPUT_FILE, importSource: 'react-router-dom' });
    const generatedCode = fs.readFileSync(OUTPUT_FILE, 'utf-8');
    
    expect(generatedCode).toMatchSnapshot();
    const nestedLayoutRegex = /<Route path="dashboard" element={<Page1 \/>}>\s*<Route index element={<Page2 \/>} \/>\s*<Route path="settings" element={<Page3 \/>} \/>\s*<\/Route>/s;
    expect(generatedCode).toContain('<Route path="/" element={<Page0 />}>');
    expect(generatedCode).toMatch(nestedLayoutRegex);
  });

  it('should handle pages alongside a directory with a layout', async () => {
    mockedGlob.mockResolvedValue([
      `${PAGES_DIR}/admin/layout.tsx`,
      `${PAGES_DIR}/admin/index.tsx`,
      `${PAGES_DIR}/admin/users.tsx`,
      `${PAGES_DIR}/index.tsx`,
    ]);
    vol.fromJSON({
      'admin/layout.tsx': 'export default () => <Outlet />', // Page0
      'admin/index.tsx': 'export default () => <div>Admin</div>', // Page1
      'admin/users.tsx': 'export default () => <div>Admin Users</div>', // Page2
      'index.tsx': 'export default () => <div>Home</div>', // Page3
    }, PAGES_DIR);
    
    await generateRoutes({ pagesDir: PAGES_DIR, outputFile: OUTPUT_FILE, importSource: 'react-router-dom' });
    const generatedCode = fs.readFileSync(OUTPUT_FILE, 'utf-8');

    expect(generatedCode).toMatchSnapshot();
    expect(generatedCode).toContain('<Route index element={<Page3 />} />');
    const adminRegex = /<Route path="admin" element={<Page0 \/>}>\s*<Route index element={<Page1 \/>} \/>\s*<Route path="users" element={<Page2 \/>} \/>\s*<\/Route>/s;
    expect(generatedCode).toMatch(adminRegex);
  });
});