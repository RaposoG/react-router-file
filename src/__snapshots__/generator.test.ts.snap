// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateRoutes() > should create a correct snapshot for a complex structure 1`] = `
"// ATTENTION: This file is automatically generated by vite-react-file-router.
// Do not edit manually.
import React, { lazy, Suspense } from 'react';
// 👇 3. USE A VARIÁVEL PARA O IMPORT DINÂMICO AQUI
import { Routes, Route } from 'react-router-dom';

const Page0 = lazy(() => import('./pages/index.tsx'));
const Page1 = lazy(() => import('./pages/about.tsx'));
const Page2 = lazy(() => import('./pages/users/index.tsx'));
const Page3 = lazy(() => import('./pages/users/[id].tsx'));
const Page4 = lazy(() => import('./pages/users/[id]/settings.tsx'));

const LoadingComponent = () => null;

export const AppRoutes = () => (
  <Suspense fallback={<LoadingComponent />}>
    <Routes>
      <Route path="/users/:id/settings" element={<Page4 />} />
      <Route path="/users/:id" element={<Page3 />} />
      <Route path="/about" element={<Page1 />} />
      <Route path="/users" element={<Page2 />} />
      <Route path="/" element={<Page0 />} />
    </Routes>
  </Suspense>
);"
`;
